File: android/src/main/java/com/rn/s/baidumap/modules/OfflineMapModule.java
package com.rn.s.baidumap.modules;

import android.util.Log;

import com.baidu.mapapi.map.offline.MKOLSearchRecord;
import com.baidu.mapapi.map.offline.MKOLUpdateElement;
import com.baidu.mapapi.map.offline.MKOfflineMap;
import com.baidu.mapapi.map.offline.MKOfflineMapListener;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;

import java.util.ArrayList;

import androidx.annotation.NonNull;

public class OfflineMapModule extends ReactContextBaseJavaModule implements MKOfflineMapListener {

    private Promise globalPromise;
    private MKOfflineMap mOffline = null;

    public OfflineMapModule(ReactApplicationContext reactContext) {
        super(reactContext);
        mOffline = new MKOfflineMap();
        mOffline.init(this);
    }

    @NonNull
    @Override
    public String getName() {
        return "OfflineMapModule";
    }

    @ReactMethod
    public void start(int cityId) {
        mOffline.start(cityId);
    }

    @ReactMethod
    public void update(int cityId) {
        mOffline.update(cityId);
    }

    @ReactMethod
    public void stop(int cityId) {
        mOffline.pause(cityId);
    }

    @ReactMethod
    public void remove(int cityId) {
        mOffline.remove(cityId);
    }

    @ReactMethod
    public void getHotCityList(Promise promise) {
        globalPromise = promise;
        // 获取热闹城市列表
        ArrayList<MKOLSearchRecord> records = mOffline.getHotCityList();
        WritableArray array = Arguments.createArray();
        if (records != null) {
            array = this.groupCityList(records);
        }
        groupResult(array);
    }

    @ReactMethod
    public void getOfflineAllCityList(Promise promise) {
        globalPromise = promise;
        // 获取热闹城市列表
        ArrayList<MKOLSearchRecord> records = mOffline.getOfflineCityList();
        WritableArray array = Arguments.createArray();
        if (records != null) {
            array = this.groupCityList(records);
        }
        groupResult(array);
    }

    @ReactMethod
    public void searchCity(String city, Promise promise) {
        globalPromise = promise;
        // 获取热闹城市列表
        ArrayList<MKOLSearchRecord> records = mOffline.searchCity(city);
        if (records != null) {
            groupResult(this.groupCityList(records));
        }else{
            promise.reject("-1", city+"不支持城市离线地图");
        }
    }

    @ReactMethod
    public void getDownloadedCityList(Promise promise) {
        globalPromise = promise;
        // 获取热闹城市列表
        ArrayList<MKOLUpdateElement> records = mOffline.getAllUpdateInfo();
        WritableArray array = Arguments.createArray();
        if (records != null) {
            for (MKOLUpdateElement item : records) {
                WritableMap writableMap = Arguments.createMap();
                writableMap.putInt("cityID", item.cityID);
                writableMap.putString("cityName", item.cityName);
                writableMap.putInt("ratio", item.ratio);
                writableMap.putInt("status", item.status);
                writableMap.putBoolean("update", item.update);
                writableMap.putString("size", this.formatDataSize(item.size));
                writableMap.putString("serversize", this.formatDataSize(item.serversize));
                writableMap.putInt("level", item.level);
                writableMap.putDouble("longitude", item.geoPt.longitude);
                writableMap.putDouble("latitude", item.geoPt.latitude);
                array.pushMap(writableMap);
            }
        }
        groupResult(array);
    }

    @Override
    public void onGetOfflineMapState(int type, int state) {
        switch (type) {
            case MKOfflineMap.TYPE_DOWNLOAD_UPDATE: {
                MKOLUpdateElement update = mOffline.getUpdateInfo(state);
                Log.i("test", "TYPE_DOWNLOAD_UPDATE");
                // 处理下载进度更新提示
                if (update != null) {
                    Log.d("OfflineDemo", String.format("%s : %d%%", update.cityName, update.ratio));
//                    isDownLoadedOfflineMap = true;
                }
            }
            break;
            case MKOfflineMap.TYPE_NEW_OFFLINE:
                Log.i("test", "TYPE_NEW_OFFLINE");
                // 有新离线地图安装
                break;
            case MKOfflineMap.TYPE_VER_UPDATE:
                Log.i("test", "TYPE_VER_UPDATE");
                // 版本更新提示
                // MKOLUpdateElement e = mOffline.getUpdateInfo(state);
                break;
            default:
                break;
        }
    }

    public WritableArray groupCityList(ArrayList<MKOLSearchRecord> records) {
        WritableArray array = Arguments.createArray();
        for (MKOLSearchRecord item : records) {
            WritableMap writableMap = Arguments.createMap();
            int cityType = item.cityType;
            writableMap.putInt("cityType", cityType);
            writableMap.putInt("cityID", item.cityID);
            writableMap.putString("cityName", item.cityName);
            writableMap.putString("dataSize", this.formatDataSize(item.dataSize));
            if(cityType == 1) {
                writableMap.putArray("childCities", groupCityList(item.childCities));
            }
            array.pushMap(writableMap);
        }
        return array;
    }

    public String formatDataSize(long size) {
        String ret = "";
        if (size < (1024 * 1024)) {
            ret = String.format("%dK", size / 1024);
        } else {
            ret = String.format("%.1fM", size / (1024 * 1024.0));
        }
        return ret;
    }

    public void groupResult(WritableArray writableArray) {
        WritableMap writableMap = Arguments.createMap();
        writableMap.putArray("list", writableArray);
        writableMap.putInt("code", 0);
        globalPromise.resolve(writableMap);
    }

}

File: iosLib/RNSBaidumap/mapModules/OfflineMapModule.swift
//
//  OfflineMapModule.swift
//  reactNativeSBaidumap
//
//  Created by Arno on 2020/1/5.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation

@objc(OfflineMapModule)
class OfflineMapModule: NSObject, BMKOfflineMapDelegate {
  
  let kilobyte = 1024
  let megabyte = 1048576
  let gigabyte = 1073741824
  var prmResolve: RCTPromiseResolveBlock!
  var prmReject: RCTPromiseRejectBlock!
  //离线地图类的实例
  lazy var offlineMap: BMKOfflineMap = {
    let newOfflineMap = BMKOfflineMap()
    //设置离线地图类的代理
    newOfflineMap.delegate = self
    return newOfflineMap
  }()
  
  
  @objc func start(_ cityId: NSInteger) {
    offlineMap.start(Int32(cityId))
  }
  
  @objc
  func update(_ cityId: NSInteger) {
    offlineMap.update(Int32(cityId))
  }
  
  @objc
  func stop(_ cityId: NSInteger) {
    offlineMap.pause(Int32(cityId))
  }
  
  @objc
  func remove(_ cityId: NSInteger) {
    offlineMap.remove(Int32(cityId))
  }
  
  @objc
  func getHotCityList(_ resolve:@escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
    prmResolve = resolve;
    prmReject = reject;
    let hotCitys = offlineMap.getHotCityList() as! [BMKOLSearchRecord]
    self.groupResult(list: self.groupCityList(records: hotCitys));
  }
  
  @objc
  func getOfflineAllCityList(_ resolve:@escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
    prmResolve = resolve;
    prmReject = reject;
    let list = offlineMap.getOfflineCityList() as! [BMKOLSearchRecord]
    self.groupResult(list: self.groupCityList(records: list));
  }
  
  @objc
  func searchCity(_ city: String, _ resolve:@escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
    prmResolve = resolve;
    prmReject = reject;
    let list = offlineMap.searchCity(city) as! [BMKOLSearchRecord]
    self.groupResult(list: self.groupCityList(records: list));
  }
  
  @objc
  func getDownloadedCityList(_ resolve:@escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
    prmResolve = resolve;
    prmReject = reject;
    let records = offlineMap.getAllUpdateInfo()
    var list = Array<Dictionary<String, Any>>();
    if(records != nil) {
      let newRecords = records as! [BMKOLUpdateElement]
      for item in newRecords {
        var data = Dictionary<String, Any>();
        data["cityID"] = item.cityID;
        data["cityName"] = item.cityName;
        data["ratio"] = item.ratio;
        data["status"] = item.status;
        data["update"] = item.update;
        data["size"] = self.getDataSizeString(NSInteger(item.size));
        data["serversize"] = self.getDataSizeString(NSInteger(item.serversize));
        data["latitude"] = item.pt.latitude;
        data["longitude"] = item.pt.longitude;
        list.append(data)
      }
    }
    self.groupResult(list: list);
  }
  
  /**
  返回通知结果
  @param type 事件类型
  @param state 事件状态
  type为TYPE_OFFLINE_UPDATE，表示正在下载或更新城市id为state的离线包，
  type为TYPE_OFFLINE_ZIPCNT，表示检测到state个离线压缩包，
  type为TYPE_OFFLINE_ADD，表示新安装的离线地图数目，
  type为TYPE_OFFLINE_UNZIP，表示正在解压第state个离线包，
  type为TYPE_OFFLINE_ERRZIP，表示有state个错误包，
  type为TYPE_VER_NEWVER，表示id为state的城市离线包有新版本，
  type为TYPE_OFFLINE_UNZIPFINISH时，表示扫瞄完成，成功导入state个离线包
  */
  func onGetOfflineMapState(_ type: Int32, withState state: Int32) {
    print("onGetOfflineMapState")
  }
  
  /**
   离线地图数据包大小单位转换
   @param packetSize 离线地图数据包总大小，单位是bytes
   @return 转换单位后的离线地图数据包大小
   */
  func getDataSizeString(_ packetSize: NSInteger) -> String {
      var packetSizeString: String = ""
      if packetSize < kilobyte {
          packetSizeString = String(format: "%ldB", packetSize)
      } else if packetSize < megabyte {
          packetSizeString = String(format: "%ldK", packetSize / kilobyte)
      } else if packetSize < gigabyte {
          if (packetSize % megabyte) == 0 {
          packetSizeString = String(format: "%ldM", megabyte)
          } else {
              var decimal: NSInteger = 0
              var decimalString: String = ""
              decimal = packetSize % megabyte
              decimal /= kilobyte
              if decimal < 10 {
              decimalString = String(format: "%d", 0)
              } else if decimal >= 10 && decimal < 100 {
                  let temp: NSInteger = decimal / 10
                  if temp >= 5 {
                      decimalString = String(format: "%d", 1)
                  } else {
                      decimalString = String(format: "%d", 0)
                  }
              } else if decimal >= 100 && decimal < kilobyte {
                  let temp: NSInteger = decimal / 100
                  if temp >= 5 {
                      decimal = temp + 1
                      if decimal >= 10 {
                          decimal = 9
                      }
                      decimalString = String(format: "%ld", decimal)
                  } else {
                      decimalString = String(format: "%ld", temp)
                  }
              }
              if decimalString.isEmpty {
                  packetSizeString = String(format: "%ldMss", packetSize / megabyte)
              } else {
                  packetSizeString = String(format: "%ld.%@M", packetSize / megabyte, decimalString)
              }
          }
      } else {
          packetSizeString = String(format: "%ldG", packetSize / gigabyte)
      }
      return packetSizeString
  }
  
  func groupCityList(records: [BMKOLSearchRecord]) -> Array<Dictionary<String, Any>> {
    var list = Array<Dictionary<String, Any>>();
    for item in records {
      var data = Dictionary<String, Any>();
      let cityType = item.cityType;
      data["cityType"] = cityType;
      data["cityID"] = item.cityID;
      data["cityName"] = item.cityName;
      data["dataSize"] = self.getDataSizeString(NSInteger(item.size));
      if(cityType == 1) {
        data["childCities"] = self.groupCityList(records: item.childCities as! [BMKOLSearchRecord]);
      }
      list.append(data)
    }
    return list;
  }
  
  public func groupResult(list: Array<Dictionary<String, Any>>) {
    var data = Dictionary<String, Any>();
    data["list"] = list;
    data["code"] = 0;
    prmResolve(data);
  }
  
}
